### ISSUE TITLE
[Bug]: Getting "A router only supports one blocker at a time" warning during HMR when running your navigation-blocking example

### ISSUE THREAD

**Issue by simonbaas** [remix-run/react-router#11430]
### What version of React Router are you using?

6.22.3

### Steps to Reproduce

1. Clone react-router repo and navigate into examples/navigation-blocking
2. Update react-router-dom to 6.22.3 (same issue on ^6.21.0 though)
3. Run npm install
4. Start dev server: npm run dev
5. On page go to http://localhost:3000/three (Three (Form with blocker)
6. In code editor, press save on app.tsx so vite HMR runs
7. Go back to web page and click on another link
8. Notice warning in console

### Expected Behavior

Not getting the warning. Any blocker should be disposed/cleaned up correctly.

### Actual Behavior

Getting warning "A router only supports one blocker at a time" in console.

I'm seeing this issue in real world case as well, even if useBlocker is only called once in the entire app. It the component which calls useBlocker is "mounted" (in lack of better term) and the router is re-created higher up it seems like the old blocker is not disposed/removed/cleanup correctly.

**Comment by elvince** [comment]
Hi,
I'm facing exactly this bug also.

If a new router is recreated while a useBlocker is "active/mounted", it is not reseted properly.

Thanks for your help in fixing this.

**Comment by JasonDyke** [comment]
I've been seeing this same issue for a while now, but after updating to react-router-dom v6.23.0 today I haven't seen it again.

Can anyone else confirm that this issue is fixed in 6.23.0?

**Comment by Jose4gg** [comment]
@JasonDyke it is still happening to me in `6.23.0`

**Comment by simonbaas** [comment]
Same issue still exists on 6.23.0.

**Comment by kubmir** [comment]
Same issue for me on 6.23.0. In fact, it also really prevents navigation for me. The last version which worker for me was 6.13.0 - all versions since then breaks navigation after form submission for me.

**Comment by elvince** [comment]
Same issue still exists on 6.23.0.

**Comment by ptim** [comment]
Not sure if this is a valid solution, but I notice in the example:
https://github.com/remix-run/react-router/blob/main/examples/navigation-blocking/src/app.tsx#L42

```tsx
if (import.meta.hot) {
  import.meta.hot.dispose(() => router.dispose());
}
```

...seems to resolve the warning for me

**Comment by simonbaas** [comment]
I see that the import.meta.hot.dispose(...) function is called in the example, but it doesn't solve the issue because you still get the "A router only supports one blocker at a time" message in the browser dev console if you follow the steps mentioned in the original post above. (Just re-tested it with version 6.24.0).

Also, this could only be a potential solution for when in development mode and because the issue also is present when "a new router is recreated while a useBlocker is "active/mounted"" (as mentioned above as well) it's not a valid solution imho.

**Comment by samuelblattner** [comment]
We have the same issue, because our routes depend on a state and recreate the router whenever this state changes. As a temporary fix, we create the router only when the module is loaded and pass route updates via ``_internalSetRoutes``, which clearly is a hack. But it seems to do the job for now...

**Comment by abentele** [comment]
same problem here, with react-router 6.26.1: if I change something in code (with hot-reload in Browser) while a useBlocker is mounted, the warning "A router only supports one blocker at a time" is printed.

=> the conclusion of @simonbaas seems to be the most reasonable related to my issue.

**Comment by KarineBrandelli** [comment]
Still happening at v7.

**Comment by andrecarrblad** [comment]
I only experience this issue with React 19. I have tested the following versions:

React 18 together with react-router-dom 6.28.0 – works fine

React 18 together with react-router-dom 7.1.1 – works fine

React 19 together with react-router-dom 7.1.1 – works fine

React 19 together with react-router-dom 6.28.0 – I get warning: A router only supports one blocker at a time.

**Comment by Rey-Wang** [comment]
@WaheedRumaneh when the issue happened, did the router freeze?

**Comment by JasonDyke** [comment]
I see the locking.

Currently using react-router-dom@7.2.0, react-router@7.2.0.

**Comment by KagaKyouichirou** [comment]
```
  // This effect is in charge of blocker key assignment and deletion (which is
  // tightly coupled to the key)
  React.useEffect(() => {
    let key = String(++blockerId);
    setBlockerKey(key);
    return () => router.deleteBlocker(key);
  }, [router]);

  // This effect handles assigning the blockerFunction.  This is to handle
  // unstable blocker function identities, and happens only after the prior
  // effect so we don't get an orphaned blockerFunction in the router with a
  // key of "".  Until then we just have the IDLE_BLOCKER.
  React.useEffect(() => {
    if (blockerKey !== "") {
      router.getBlocker(blockerKey, blockerFunction);
    }
  }, [router, blockerKey, blockerFunction]);
```

This is from `useBlocker` (see [hooks.tsx](https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/hooks.tsx)) 
When `router` changes, the first effect runs the cleanup and the old key ('42' for example) is removed from a global map `blockerFunctions` (see [deleteBlocker](https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/router/router.ts#L3033)), and then it calls `setBlockerKey(key)` with the new key ('43'). But then the second effect runs, and the `blockerKey` it captures is still the **old version** (because it's not the new re-render cycle yet), which is '42'; and `router.getBlocker` ([getBlocker](https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/router/router.ts#L3033)) would query and set the fn into the `blockerFunctions` with the '42' key and it would stay in the map. Then the second effect runs another time due to the newly set `blockerKey`.
So `blockerFunctions` would contain more than one function and the warning would appear.

I think that's how the blocker's state turn ill.

**Comment by pbatey** [comment]
I found creating a `useEffect` that calls `blocker.reset()` in a cleanup function solved my problem with the "A router only supports one blocker at a time" warning.

```
const BlockNavigation = ({enabled = false}) => {
  const blocker = useBlocker(({ currentLocation, nextLocation }) => {
    const navAway = currentLocation.pathname !== nextLocation.pathname;
    return navAway && enabled;
  });

  useEffect(() => {
    // ensure the blocker is reset at unmount
    return () => {
      if (blocker.state == 'blocked') blocker.reset();
    };
  }, [blocker]);

  // Handle confirm navigation
  const handleConfirm = () => {
    if (blocker.state === 'blocked') blocker.proceed();
  };

  // Handle cancel navigation
  const handleCancel = () => {
    if (blocker.state === 'blocked') blocker.reset();
  };

  return <Modal ...
```

**Comment by Haschtl** [comment]
I fixed the warning by using a Context for the Blocker
This snippet also improves the behavior if multiple blocker are registered. It doesn't proceed if multiple blocker are active.

```typescript
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useRef,
} from "react";
import { type Blocker, type BlockerFunction, useBlocker } from "react-router";

const BlockerContext = createContext<{
  blocker: Blocker;
  register: (fn: BlockerFunction) => () => void;
}>({
  blocker: {},
  register: () => () => { },
});

export const BlockerProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const blockersRef = useRef<Set<BlockerFunction>>(new Set());

  const register = useCallback((fn: BlockerFunction) => {
    blockersRef.current.add(fn);
    return () => blockersRef.current.delete(fn);
  }, []);

  const shouldBlock = useCallback((...v: Parameters<BlockerFunction>) => {
    for (const fn of blockersRef.current) {
      if (fn(...v)) {
        return true;
      }
    }
    return false;
  }, []);

  const blocker = useBlocker(shouldBlock);
  const proceed = () => {
    if (blockersRef.current.size <= 1) blocker.proceed?.();
  };
  return (
    <BlockerContext.Provider
      value={{ blocker: { ...blocker, proceed }, register }}
    >
      {children}
    </BlockerContext.Provider>
  );
};

export const useRegisterBlocker = (fn: BlockerFunction, enabled = true) => {
  const { register, blocker } = useContext(BlockerContext);

  useEffect(() => {
    if (!enabled) return;
    return register(fn);
  }, [register, fn, enabled]);
  return blocker;
};

```

I placed the blocker-provider directly under the router-provider

**Comment by broodfusion** [comment]
> I have same problem. After hot reload while in page with active `useBlocker`, the links stop working and I get warning message in console "A router only supports one blocker at a time".
> 
> What I figured out is that my `shouldBlock = useCallback` function is cached with some "blocked" state that tells the blocker to block. So as a workaround I tried to add a ref which worked:
> 
>   const isMountedRef = useRef<boolean>();
>   useEffect(() => {
>     isMountedRef.current = true;
>     return () => {
>       isMountedRef.current = false;
>     };
>   }, []);
> 
>   const shouldBlock = useCallback<BlockerFunction>(
>     ({ currentLocation, nextLocation }) => {
>       return (
>         isMountedRef.current &&
>         currentLocation.pathname !== nextLocation.pathname
>         // your other blocking conditions here
>       );
>     },
>     [state],
>   );
> 
>   const blocker = useBlocker(shouldBlock);
> That doesn't solve the warning though

Using the cleanup function in the useEffect inadvertently solves my problem with useBlocker not working well with the browser's back/forward. This is version 7.5.1

**Comment by hammadfauz** [comment]
> To align with our new [Open Governance](https://remix.run/blog/rr-governance) model, we are now requiring that all issues have a [**minimal** and **runnable** reproduction](https://github.com/remix-run/react-router/blob/main/GOVERNANCE.md#bugissue-process). To that end, we're doing some housekeeping in the repo to clean up existing issues that do not have a valid reproduction. This should get us down to a more manageable number of issues and allow us to be more responsive to existing and newly filed issues.
> 
> We're using a Github actions script to identify issues without a reproduction by looking for a [StackBlitz](https://stackblitz.com/), [CodeSandbox](https://codesandbox.io/), or [Github](https://github.com) link in the issue body. This won't be perfect, so if this issue has a reproduction on another platform, please comment back on here and we can re-open the issue. Similarly, if there's a reproduction buried in a comment, please move the link into the description and comment back. Please tag `@brophdawg11` or `@brookslybrand` in your comment so we get a notification as well 🙂.
> 
> If this issue did not have a reproduction but is still valid, or if you wish to start with a fresh issue, please [create a new issue](https://github.com/remix-run/react-router/issues/new?template=bug_report.yml) with a fresh reproduction against v7 and link to this issue in the new description.
> 
> If this a feature request, please open a new [Proposal Discussion](https://github.com/remix-run/react-router/discussions/new?category=proposals) in React Router, and if it gets enough community support it can be considered for implementation.
> 
> If you have any questions you can always reach out on [Discord](https://rmx.as/discord). Thanks again for providing feedback and helping us make React Router even better!

This issue has reproduction steps.

**Comment by gerpres** [comment]
> I also got this warning when i try to use a reactive router object (not defined outside the react component tree).
> 
> ```
>   // This effect is in charge of blocker key assignment and deletion (which is
>   // tightly coupled to the key)
>   React.useEffect(() => {
>     let key = String(++blockerId);
>     setBlockerKey(key);
>     return () => router.deleteBlocker(key);
>   }, [router]);
> 
>   // This effect handles assigning the blockerFunction.  This is to handle
>   // unstable blocker function identities, and happens only after the prior
>   // effect so we don't get an orphaned blockerFunction in the router with a
>   // key of "".  Until then we just have the IDLE_BLOCKER.
>   React.useEffect(() => {
>     if (blockerKey !== "") {
>       router.getBlocker(blockerKey, blockerFunction);
>     }
>   }, [router, blockerKey, blockerFunction]);
> ```
> 
> This is from `useBlocker` (see [hooks.tsx](https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/hooks.tsx#L1256)) When `router` changes, the first effect runs the cleanup and the old key ('42' for example) is removed from a global map `blockerFunctions` (see [deleteBlocker](https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/router/router.ts#L3033)), and then it calls `setBlockerKey(key)` with the new key ('43'). But then the second effect runs, and the `blockerKey` it captures is still the **old version** (because it's not the new re-render cycle yet), which is '42'; and `router.getBlocker` ([getBlocker](https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/router/router.ts#L3033)) would query and set the fn into `blockerFunctions` with the '42' key and it would stay in the map. Then the second effect runs another time due to the newly set `blockerKey`. So `blockerFunctions` would contain more than one function and the warning would appear.
> 
> I think that's how the blocker's state turns ill.

also see the same warning, tracked it down to exactly the cause described above.
any plans to correct cleanup-logic?
